[Merged by Bors] - feat(algebra/lie/nilpotent): basic lemmas about nilpotency for Lie subalgebras of associative algebras

ocfnash opened the pull request:
The main lemma is: `lie_algebra.is_nilpotent_ad_of_is_nilpotent` which is the first step in proving Engel's theorem.

---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/lie/nilpotent): basic lemmas about nilpotency for Lie subalgebras of associative algebras

The main lemma is: `lie_algebra.is_nilpotent_ad_of_is_nilpotent` which is the first step
in proving Engel's theorem.

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1195,6 +1199,51 @@ by { ext, simp only [linear_map.id_coe, mul_one, id.def, lmul_right_apply] }
 
+@[simp] lemma lmul_left_zero_eq_zero :
+  lmul_left R (0 : A) = 0 :=
+by { ext, simp, }
```
Not shorter, but gives a stronger hint "we already proved this"
```suggestion
(lmul R A).map_zero
```

ocfnash replied:
Thanks!

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1195,6 +1199,51 @@ by { ext, simp only [linear_map.id_coe, mul_one, id.def, lmul_right_apply] }
+  { simp only [linear_map.id_coe, linear_map.one_apply, lmul_left_one, id.def, pow_zero], },
+  { simp only [lmul_left_apply] at ih,
+    simp only [lmul_left_apply, pow_succ, linear_map.mul_apply, ih, mul_assoc], },
+end
```
```suggestion
((lmul R A).map_pow a n).symm
```

ocfnash replied:
Thanks!

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1195,6 +1199,51 @@ by { ext, simp only [linear_map.id_coe, mul_one, id.def, lmul_right_apply] }
+  { rw lmul_right_apply at ih,
+    rw [pow_succ, pow_succ, linear_map.mul_apply, ih, lmul_right_apply, lmul_right_apply,
+      mul_assoc, ← pow_succ, ← pow_succ'], },
+end
```
Slight shorter with an import of `algebra.iterate_hom`:
```suggestion
begin
  apply linear_map.coe_injective,
  dsimp only, -- TODO[gh-8102]: remove this once `coe_injective` does not introduce junk
  rw linear_map.coe_pow,
  exact mul_right_iterate a n,
end
```
(xref gh-8102)

ocfnash replied:
Nice!

eric-wieser committed:
fix(algebra/module/linear_map): do not introduce `show`

Without this change, `apply linear_map.coe_injective` introduces some unpleasant `show` terms

jcommelin commented on src/algebra/algebra/basic.lean:
```diff
@@ -1195,6 +1199,51 @@ by { ext, simp only [linear_map.id_coe, mul_one, id.def, lmul_right_apply] }
+end
+
+@[simp] lemma pow_lmul_right (a : A) (n : ℕ) :
+  (lmul_right R a)^n = lmul_right R (a ^n) :=
```
```suggestion
  (lmul_right R a) ^ n = lmul_right R (a ^ n) :=
```

jcommelin reviewed (commented):
minor comment. apart from Eric's comments, lgtm

ocfnash committed:
Suggestions from CR

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1224,24 +1225,15 @@ begin
-    rw [pow_succ, pow_succ, linear_map.mul_apply, ih, lmul_right_apply, lmul_right_apply,
-      mul_assoc, ← pow_succ, ← pow_succ'], },
+  apply linear_map.coe_injective,
+  dsimp only, -- TODO[gh-8102]: remove this once `coe_injective` does not introduce junk
```
Are you happy to wait for this to be merged first?

ocfnash replied:
Absolutely!

How about I cherry pick its commit and make it a dependency and then shorten this proof?

ocfnash replied:
I've actually just done this.

ocfnash committed:
Golf proof using #8102
